File: 백준/Silver/2805. 나무 자르기/나무 자르기.js
const filePath = process.platform === 'linux' ? '/dev/stdin' : 'e.txt';
let input = require('fs').readFileSync(filePath).toString().trim().split('\n');
const [treeCount, need] = input.shift().split(' ').map(Number);

input = input[0].split(' ').map(Number);
/*
나무의 길이가 20억 이상이므로 시간초과에 조심해야한다.

원래라면 0부터 max나무길이까지를 반복으로 돌아서 구해야한다.
이런 반복을 이분 탐색으로 할 수 있다.
*/

let maxLen = -Infinity;

const binarySearch = (start, end) => {
	const mid = Math.floor((start + end) / 2);
	if (start > end) return;

	const treeLength = input.reduce((sum, cur) => {
		if (cur - mid > 0) return sum + (cur - mid);
		else return sum;
	}, 0);

	if (treeLength >= need) {
		if (maxLen < mid) maxLen = mid;
		binarySearch(mid + 1, end);
	}
	if (treeLength < need) {
		binarySearch(start, mid - 1);
	}
};

binarySearch(0, Math.max(...input));

console.log(maxLen);

File: 백준/Silver/2805. 나무 자르기/나무 자르기.py
'''
하나의 변수 x에 대해서 문제의 답이 FFFFFFTTTT처럼 변화가 존재하는 경우 이진 탐색으로 풀 수 있다.

여기선 톱의 톱이 h에 대해서 문제의 답이 FFFFFFTTT처럼 변한다.

근데 이건 TTTTTTFF가 된다. 따라서 첫 T를 찾으려면 lo를 출력해야한다. 

이때의 결정 문제: getTreeHeight(m) >= x인가?

그럼 이 경우 시간복잡도는 어떻게 되는가? logN * n이다
이때의 n은 1,000,000,000
'''
import math

n, m = map(int, input().split())
trees = list(map(int, input().split()))

def getTreeHeight(h):
    sum = 0
    
    for i in range(n):
        sum += 0 if trees[i] - h <= 0 else trees[i] - h
    
    return sum
    
def binarySearch(x):
    lo = -1
    hi = 1000000001
    
    while lo + 1 < hi: # 사이에 한 칸도 여유가 없을 때 종료된다. 
        m = math.trunc((hi + lo) / 2)
        if getTreeHeight(m) < x: # 더 많이 잘라야 해
            hi = m
        else:
            lo = m
        
    return lo # TTTTTFFF로 변하는 문제이기 때문에 lo를 출력해야한다.

print(binarySearch(m))


